File: i/image-overlap.rs
impl Solution {
  pub fn largest_overlap(img1: Vec<Vec<i32>>, img2: Vec<Vec<i32>>) -> i32 {
    let n = img1.len();
    let mut ans = 0;
    for dx in 0..2 * n { for dy in 0..2 * n {
      let mut cur = 0;
      for i in 0..n { for j in 0..n {
        let (x, y) = (i + dx - n, j + dy - n);
        if 0 <= x && x < n && 0 <= y && y < n && img1[i][j] == 1 && img1[i][j] == img2[x][y] {
          cur += 1;
        }
      }}
      ans = ans.max(cur);
    }}
    ans
  }
}

File: i/image-overlap.cc
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
  int largestOverlap(vector<vector<int>>& img1, vector<vector<int>>& img2) {
    int N = img1.size(), ans = 0;
    auto oob = [&](int i, int j) {
      return i < 0 || i >= N || j < 0 || j >= N;
    };
    for (int i2 = -N; i2 < N; ++i2) for (int j2 = -N; j2 < N; ++j2) {
      int cnt = 0;
      for (int i = 0; i < N; ++i) for (int j = 0; j < N; ++j) {
        if (oob(i + i2, j + j2)) continue;
        cnt += img1[i][j] & img2[i + i2][j + j2];
      }
      ans = max(ans, cnt);
    }
    return ans;
  }
};
